crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -191,46 +228,499 @@ pub fn run_enter_schedule<S: States>(world: &mut World) {
-                    world.send_event(StateTransitionEvent {
-                        before: exited.clone(),
-                        after: entered.clone(),
-                    });
-                    // Try to run the schedules if they exist.
-                    world.try_run_schedule(OnExit(exited.clone())).ok();
-                    world
-                        .try_run_schedule(OnTransition {
-                            from: exited,
-                            to: entered.clone(),
-                        })
-                        .ok();
-                    world.try_run_schedule(OnEnter(entered)).ok();
-                }
+
+    match next_state_resource {
+        NextState::Set(new_state) => {
+            let new_state = new_state.clone();
+            internal_apply_state_transition(world, Some(new_state));
+        }
+        NextState::Remove => {
+            internal_apply_state_transition::<S>(world, None);
+        }
+        _ => {
+            return;
+        }
+    }
+
+    world.insert_resource(NextState::<S>::Unchanged);
+}
+
+/// Trait defining a state that is automatically derived from other [`States`].
+///
+/// A Computed State is a state that is deterministically derived from a set of `SourceStates`.
+/// The [`StateSet`] is passed into the `compute` method whenever one of them changes, and the
+/// result becomes the state's value.
+///
+/// ```
+/// /// Computed States require some state to derive from
+/// #[derive(States, Clone, PartialEq, Eq, Hash, Debug, Default)]
+/// enum AppState {
+///     #[default]
+///     Menu,
+///     InGame { paused: bool }
+/// }
+///
+///
+/// #[derive(Clone, PartialEq, Eq, Hash, Debug)]
+/// struct InGame;
+///
+/// impl ComputedStates for InGame {
+///     /// We set the source state to be the state, or set of states,
+///     /// we want to depend on.
+///     type SourceStates = AppState;
+///
+///     /// We then define the compute function, which takes in
+///     /// either a single optional state, or a tuple of optional
+///     /// states based on whether our source is one state
+///     /// or many.
+///     fn compute(sources: Option<AppState>) -> Option<Self> {
+///         match sources {
+///             /// When we are in game, we want to return the InGame state
+///             Some(AppState::InGame { .. }) => Some(InGame)
+///             /// Otherwise, we don't want the `State<InGame>` resource to exist,
+///             /// so we return None.
+///             _ => None
+///         }
+///     }
+/// }
+///
+/// # struct App;
+/// # impl App {
+/// #   fn new() -> Self { App }
+/// #   fn init_state<S: States>(&mut self) -> &mut Self {self}
+/// #   fn add_computed_state<S: ComputedStates>(&mut self) -> &mut Self {self}
+/// # }
+///
+/// /// you can then add it to an App, and from there you use the state as normal
+///     App::new()
+///         .init_state::<AppState>()
+///         .add_computed_state::<InGame>()
+/// ///...
```

Comment:
Is this line intentional?

Reply:
The `/// ...` was intentional, but if it doesn't match the expected style I can remove it.

Reply:
Oh I see what you're intending now. Yeah, let's remove it for style.